File: worker_py/agents/lesson.py
"""Lesson-generation agent. Port of ``worker/agents/lesson.ts``.

Retrieve grounding material (RAG, NotebookLM-pluggable), author a level-matched
micro-lesson, persist it under ``lesson:gen:{id}`` and index it for reuse.
"""

import json
import re
from typing import Any, Dict, List, Optional

from redis.asyncio import Redis

from ..agent import run_skill_agent
from ..bus import publish as bus_publish
from ..loops.l1 import get_level, level_prompt_hint, record_signal
from ..rag import embed, ensure_vector_index, index_material, search_materials
from ..summarize import notebooklm_summarize
from ..util import now_ms


def _lesson_key(lid: str) -> str:
    return f"lesson:gen:{lid}"


def _job_key(job_id: str) -> str:
    return f"lessonjob:{job_id}"


def _prompt(topic: str, level_hint: str, grounding: str) -> str:
    material = grounding[:6000] or "(no extra material retrieved; use well-established fundamentals)"
    return (
        f'You are Tiruno\'s lesson author. Write a short, accurate micro-lesson on "{topic}" for a learner.\n'
        f"{level_hint}\n\n"
        "Ground every claim in this source material. Do not invent facts beyond it; where it's thin, "
        "rely on well-established fundamentals of the topic:\n"
        "---\n"
        f"{material}\n"
        "---\n\n"
        "Respond with ONLY one JSON object (no prose, no code fences) of exactly this shape:\n"
        '{"title":"...","concept":"a 3-5 sentence plain-language explanation a tutor would say aloud",'
        '"questions":[{"prompt":"...","options":["..","..","..",".."],"answer":0,'
        '"explanation":"one sentence on why the correct option is right","citation":"short source label"}]}\n\n'
        'Include exactly 5 questions. Each question must have exactly 4 options and "answer" as the 0-based '
        "index of the correct option. Keep it tight and demo-sized."
    )


def _parse_json_obj(text: str, fallback: Dict[str, Any]) -> Dict[str, Any]:
    m = re.search(r"```(?:json)?\s*([\s\S]*?)```", text or "", re.IGNORECASE)
    c = m.group(1) if m else (text or "")
    start = c.find("{")
    end = c.rfind("}")
    if start == -1 or end == -1:
        return fallback
    try:
        return json.loads(c[start : end + 1])
    except Exception:  # noqa: BLE001
        return fallback


async def _gather_grounding(redis: Redis, topic: str, article_id: Optional[str] = None) -> str:
    """Build grounding context from a specific article or via RAG retrieval."""
    if article_id:
        try:
            raw = await redis.get(f"article:{article_id}")
            if raw:
                a = json.loads(raw)
                return "\n\n".join(x for x in [a.get("title"), a.get("summary"), a.get("text")] if x)
        except Exception:  # noqa: BLE001
            pass
    try:
        await ensure_vector_index(redis)
        vec = await embed(topic)
        hits = await search_materials(redis, vec, 5)
        return "\n\n".join(f"{h['title']} ({h['topic']}). {h['text']}" for h in hits)
    except Exception:  # noqa: BLE001
        return ""


async def run_generate_lesson(redis: Redis, job: Dict[str, str]) -> None:
    uid, job_id = job["uid"], job["jobId"]
    topic = (job.get("topic") or "").strip() or "this topic"
    article_id = job.get("articleId")
    lid = f"gen-{job_id}"

    async def set_job(status: str, extra: Optional[Dict[str, Any]] = None) -> None:
        await redis.set(
            _job_key(job_id),
            json.dumps({"jobId": job_id, "status": status, **(extra or {})}),
            ex=60 * 60 * 24,
        )

    await set_job("working")
    await bus_publish(redis, uid, {"jobId": job_id, "type": "progress", "status": "researching", "step": "Gathering material…"})

    grounding = await _gather_grounding(redis, topic, article_id)
    # NotebookLM grounding hook (pluggable; returns None unless enabled).
    try:
        nb = await notebooklm_summarize(grounding, topic)
        if nb:
            grounding = f"{nb}\n\n{grounding}"
    except Exception:  # noqa: BLE001
        pass

    level = await get_level(redis, uid, topic)
    await bus_publish(redis, uid, {"jobId": job_id, "type": "progress", "step": "Writing your lesson…"})

    res = await run_skill_agent(
        prompt=_prompt(topic, level_prompt_hint(level), grounding),
        skills=[],
        max_turns=6,
        on_step=lambda step: bus_publish_step(redis, uid, job_id, step),
    )

    if not res.ok:
        await set_job("error", {"error": res.error or "generation failed"})
        await bus_publish(redis, uid, {"jobId": job_id, "type": "error", "status": "error", "error": res.error or "Lesson generation failed"})
        return

    parsed = _parse_json_obj(res.text, {})
    valid = [
        q
        for q in (parsed.get("questions") or [])
        if isinstance(q, dict) and q.get("prompt") and isinstance(q.get("options"), list) and len(q["options"]) >= 2
    ]
    questions: List[Dict[str, Any]] = []
    for i, q in enumerate(valid[:5]):
        options = [str(o) for o in q["options"][:4]]
        opt_len = len(q["options"]) or 1
        try:
            ans_num = int(float(q.get("answer")))
        except (TypeError, ValueError):
            ans_num = 0
        answer = max(0, min(opt_len - 1, ans_num))
        questions.append(
            {
                "id": f"q{i + 1}",
                "prompt": str(q["prompt"]),
                "options": options,
                "answer": answer,
                "explanation": str(q.get("explanation") or ""),
                "citation": str(q.get("citation") or "Generated by Tiru"),
            }
        )

    if not questions:
        await set_job("error", {"error": "no questions produced"})
        await bus_publish(redis, uid, {"jobId": job_id, "type": "error", "status": "error", "error": "Couldn't build a lesson — try a different topic."})
        return

    lesson = {
        "id": lid,
        "title": str(parsed.get("title") or f"A lesson on {topic}"),
        "topic": topic,
        "concept": str(parsed.get("concept") or ""),
        "questions": questions,
        "generated": True,
        "createdAt": now_ms(),
        # Set when authored from an ingested article (engagement linkage).
        **({"articleId": article_id} if article_id else {}),
    }

    await redis.set(_lesson_key(lid), json.dumps(lesson), ex=60 * 60 * 24 * 7)

    # Index the new lesson so it can be recommended as "next-best material".
    try:
        vec = await embed(f"{lesson['title']}. {lesson['concept']}")
        await index_material(
            redis,
            {"id": lid, "kind": "lesson", "refId": lid, "title": lesson["title"], "topic": topic, "text": lesson["concept"]},
            vec,
        )
    except Exception:  # noqa: BLE001
        pass

    await record_signal(redis, uid, {"kind": "explore", "topic": topic})
    await set_job("ready", {"lessonId": lid, "title": lesson["title"]})
    await bus_publish(
        redis,
        uid,
        {"jobId": job_id, "type": "done", "status": "ready", "result": {"lessonId": lid, "title": lesson["title"], "topic": topic}},
    )


def bus_publish_step(redis: Redis, uid: str, job_id: str, step: str) -> None:
    """Sync ``on_step`` shim: schedule a progress publish without awaiting."""
    from ..util import fire

    fire(bus_publish(redis, uid, {"jobId": job_id, "type": "progress", "step": step}))

File: worker/agents/lesson.ts
import type Redis from "ioredis";
import { runSkillAgent } from "../lib/sdk";
import { makeBus } from "../lib/bus";
import { getLevel, levelPromptHint, recordSignal } from "../loops/l1";
import { embed } from "../../src/lib/rag/embeddings";
import { ensureVectorIndex, indexMaterial, searchMaterials } from "../../src/lib/rag/vector";
import { notebooklmSummarize } from "../../src/lib/notebooklm";

export interface GenerateLessonJob {
  uid: string;
  jobId: string;
  topic: string;
  articleId?: string;
}

interface GenMCQ {
  id: string;
  prompt: string;
  options: string[];
  answer: number;
  explanation: string;
  citation: string;
}

interface GenLesson {
  id: string;
  title: string;
  topic: string;
  concept: string;
  questions: GenMCQ[];
  generated: true;
  createdAt: number;
  /** Set when the lesson was authored from an ingested article (engagement linkage). */
  articleId?: string;
}

const lessonKey = (id: string) => `lesson:gen:${id}`;
const jobKey = (jobId: string) => `lessonjob:${jobId}`;

const PROMPT = (topic: string, levelHint: string, grounding: string) =>
  `You are Tiruno's lesson author. Write a short, accurate micro-lesson on "${topic}" for a learner.
${levelHint}

Ground every claim in this source material. Do not invent facts beyond it; where it's thin, rely on well-established fundamentals of the topic:
---
${grounding.slice(0, 6000) || "(no extra material retrieved; use well-established fundamentals)"}
---

Respond with ONLY one JSON object (no prose, no code fences) of exactly this shape:
{"title":"...","concept":"a 3-5 sentence plain-language explanation a tutor would say aloud","questions":[{"prompt":"...","options":["..","..","..",".."],"answer":0,"explanation":"one sentence on why the correct option is right","citation":"short source label"}]}

Include exactly 5 questions. Each question must have exactly 4 options and "answer" as the 0-based index of the correct option. Keep it tight and demo-sized.`;

function parseJson<T>(text: string, fallback: T): T {
  const fenced = text.match(/```(?:json)?\s*([\s\S]*?)```/i);
  const c = fenced ? fenced[1] : text;
  const start = c.indexOf("{");
  const end = c.lastIndexOf("}");
  if (start === -1 || end === -1) return fallback;
  try {
    return JSON.parse(c.slice(start, end + 1)) as T;
  } catch {
    return fallback;
  }
}

/** Build grounding context from a specific article or via RAG retrieval over the index. */
async function gatherGrounding(
  redis: Redis,
  topic: string,
  articleId?: string
): Promise<string> {
  if (articleId) {
    try {
      const raw = await redis.get(`article:${articleId}`);
      if (raw) {
        const a = JSON.parse(raw);
        return [a.title, a.summary, a.text].filter(Boolean).join("\n\n");
      }
    } catch {
      /* fall through to RAG */
    }
  }
  try {
    await ensureVectorIndex(redis);
    const vec = await embed(topic);
    const hits = await searchMaterials(redis, vec, 5);
    return hits.map((h) => `${h.title} (${h.topic}). ${h.text}`).join("\n\n");
  } catch {
    return "";
  }
}

/**
 * Lesson-generation agent: retrieve grounding material (RAG, NotebookLM-pluggable),
 * have Claude author a level-matched micro-lesson, persist it, and index it for reuse.
 */
export async function runGenerateLesson(redis: Redis, job: GenerateLessonJob): Promise<void> {
  const bus = makeBus(redis);
  const topic = job.topic.trim() || "this topic";
  const id = `gen-${job.jobId}`;

  const setJob = (status: string, extra: Record<string, unknown> = {}) =>
    redis.set(jobKey(job.jobId), JSON.stringify({ jobId: job.jobId, status, ...extra }), "EX", 60 * 60 * 24);

  await setJob("working");
  await bus.publish(job.uid, { jobId: job.jobId, type: "progress", status: "researching", step: "Gathering material…" });

  let grounding = await gatherGrounding(redis, topic, job.articleId);
  // NotebookLM grounding hook (pluggable; returns null unless NOTEBOOKLM_ENABLED=1).
  try {
    const nb = await notebooklmSummarize(grounding, topic);
    if (nb) grounding = `${nb}\n\n${grounding}`;
  } catch {
    /* best effort */
  }

  const level = await getLevel(redis, job.uid, topic);
  await bus.publish(job.uid, { jobId: job.jobId, type: "progress", step: "Writing your lesson…" });

  const res = await runSkillAgent({
    prompt: PROMPT(topic, levelPromptHint(level), grounding),
    skills: [],
    maxTurns: 6,
    onStep: (step) => void bus.publish(job.uid, { jobId: job.jobId, type: "progress", step }),
  });

  if (!res.ok) {
    await setJob("error", { error: res.error || "generation failed" });
    await bus.publish(job.uid, { jobId: job.jobId, type: "error", status: "error", error: res.error || "Lesson generation failed" });
    return;
  }

  const parsed = parseJson<{ title?: string; concept?: string; questions?: GenMCQ[] }>(res.text, {});
  const questions: GenMCQ[] = (parsed.questions ?? [])
    .filter((q) => q && q.prompt && Array.isArray(q.options) && q.options.length >= 2)
    .slice(0, 5)
    .map((q, i) => ({
      id: `q${i + 1}`,
      prompt: String(q.prompt),
      options: q.options.slice(0, 4).map((o) => String(o)),
      answer: Math.max(0, Math.min((q.options.length || 1) - 1, Number(q.answer) || 0)),
      explanation: String(q.explanation ?? ""),
      citation: String(q.citation ?? "Generated by Tiru"),
    }));

  if (!questions.length) {
    await setJob("error", { error: "no questions produced" });
    await bus.publish(job.uid, { jobId: job.jobId, type: "error", status: "error", error: "Couldn't build a lesson — try a different topic." });
    return;
  }

  const lesson: GenLesson = {
    id,
    title: String(parsed.title || `A lesson on ${topic}`),
    topic,
    concept: String(parsed.concept || ""),
    questions,
    generated: true,
    createdAt: Date.now(),
    ...(job.articleId ? { articleId: job.articleId } : {}),
  };

  await redis.set(lessonKey(id), JSON.stringify(lesson), "EX", 60 * 60 * 24 * 7);

  // Index the new lesson so it can be recommended as "next-best material".
  try {
    const vec = await embed(`${lesson.title}. ${lesson.concept}`);
    await indexMaterial(redis, { id, kind: "lesson", refId: id, title: lesson.title, topic, text: lesson.concept }, vec);
  } catch {
    /* best effort */
  }

  await recordSignal(redis, job.uid, { kind: "explore", topic });
  await setJob("ready", { lessonId: id, title: lesson.title });
  await bus.publish(job.uid, {
    jobId: job.jobId,
    type: "done",
    status: "ready",
    result: { lessonId: id, title: lesson.title, topic },
  });
}
